### ISSUE TITLE
constructors, patterns & safety

### ISSUE THREAD

**Issue by tonyday567** [santiweight/poker-base#25]
Referring to handToShape as a starting point.

```
handToShaped :: Hand -> ShapedHand
handToShaped (Hand (Card r1 s1) (Card r2 s2))
  | r1 == r2 = mkPair r1
  | s1 == s2 = unsafeMkSuited r1 r2
  | otherwise = unsafeMkOffsuit r1 r2
```

Here's two alternatives:

Guards:

```
handToShaped :: Hand -> ShapedHand
handToShaped (Hand (Card r s) (Card r' s'))
  | r == r' = MkPair r
  | s == s' && r > r' = MkSuited r r'
  | s == s' && r < r' = MkSuited r' r
  | r > r' = MkOffsuit r r'
  | otherwise = MkOffsuit r' r
```

case compare:

```
handToShaped :: Hand -> ShapedHand
handToShaped (Hand (Card r s) (Card r' s')) =
  case compare r r' of
    EQ -> MkPair r
    GT -> mk r r'
    LT -> mk r' r
  where mk = if s == s' then MkSuited else MkOffsuit
```

I think the case compare one is what you want the compiler to arrive at.

You can avoid the exception embedded in unsafeMkSuited and have the logic directly in the function body. I think this generalises across the library with some of the other unsafe functions: you have just tested the values being used to construct some data, so that the exception and guards in unsafeMkWhatever (via the underlying mkWhatever) are redundant.

I'm not a fan of the existing `MkWhatever` convention. Every constructor in haskell makes something, so it really isn't saying anything more. The usage feels much more about being unsafe, and I would prefer `UnsafeWhatever` to make that clear.

Given the usage I'm suggesting in the handToShaped examples above, I would prefer:

`data ShapedHand = Pair Rank | UnsafeOffsuit Rank Rank | UnsafeSuited Rank Rank`

Then, no pattern for Pair & bidirectional ones for Offsuit & Suited (& Hole)

```
pattern Offsuit :: Rank -> Rank -> ShapedHand
pattern Offsuit r1 r2 <- UnsafeOffsuit r1 r2
  where
    Offsuit r1 r2 = if r1 > r2 then UnsafeOffsuit r1 r2 else UnsafeOffsuit r2 r1
```

I think mkOffsuit and unsafeMkOffsuit etc are idiomatic as is.

This convention would fit in with a change to Rank to Word8-based, so we could have:

newtype Rank = UnsafeRank { rank :: Word8 }

```
pattern Rank :: Word8 -> Rank
pattern Rank r <- UnsafeRank r
  where
    Rank r = if r <0 || r > 12 then (terror "wtf") else UnsafeRank r
```

What do you think?

**Comment by santiweight** [comment]
But in many cases I see that I disambiguated unnecessarily - those cases should definitely be `Unsafe` as you say!

**Comment by santiweight** [comment]
Anywho - I do think we should address these naming issues after the full functionality fixes - which will likely be quite soon.

**Comment by santiweight** [comment]
@tonyday567 Yeah I don't love `Mk` either. Perhaps a new standard/syntax will follow namespace unification...

I can't think of a better prefix either! I think that's a good suggestion. I think Unsafe vs Mk is good. Let's go with that from now on!
